replications/internal/queue_management.go

```diff
@@ -71,44 +81,133 @@ func (qm *durableQueueManager) InitializeQueue(replicationID platform.ID, maxQue
 		return err
 	}
 
-	// Map new durable queue to its corresponding replication stream via replication ID
-	qm.replicationQueues[replicationID] = newQueue
-
 	// Open the new queue
 	if err := newQueue.Open(); err != nil {
 		return err
 	}
 
+	// Map new durable queue and scanner to its corresponding replication stream via replication ID
+	rq := replicationQueue{
+		queue:   newQueue,
+		logger:  qm.logger,
+		done:    make(chan struct{}),
+		receive: make(chan struct{}),
+	}
+	qm.replicationQueues[replicationID] = &rq // todo does this need to be held?
+	rq.Open()
+
 	qm.logger.Debug("Created new durable queue for replication stream",
 		zap.String("id", replicationID.String()), zap.String("path", dir))
 
 	return nil
 }
 
+func (rq *replicationQueue) Open() {
+	rq.wg.Add(1)
+	go rq.run()
+}
+
+func (rq *replicationQueue) Close() error {
+	close(rq.receive)
+	close(rq.done)
+	rq.wg.Wait() // wait for goroutine to finish processing all messages
+	return rq.queue.Close()
+}
+
+func (rq *replicationQueue) run() {
+	defer rq.wg.Done()
+
+	retryInterval := time.Second // todo configurable?
+	retryTimer := time.NewTicker(retryInterval)
+	defer retryTimer.Stop()
+
+	writer := func() {
+		for {
+			_, err := rq.SendWrite(func(b []byte) error {
+				rq.logger.Info("written bytes", zap.String("bytes", string(b)))
+				return nil
+			})
+			if err != nil {
+				if err == io.EOF {
+					// No more data
+					rq.logger.Debug("Finished replication writes for queue")
+				} else {
+					// todo more error handling
+					panic(1)
+				}
+				break
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-rq.done: // end the goroutine when done is messaged
+			return
+		case <-retryTimer.C: // run the scanner every 1s
+			writer()
+		case <-rq.receive: // run the scanner on data append
+			writer()
+		}
+	}
+}
+
+func (rq *replicationQueue) SendWrite(dp func([]byte) error) (int, error) {
+	// err here can be io.EOF, indicating nothing to write
+	scan, err := rq.queue.NewScanner()
+	if err != nil {
+		return 0, err
+	}
+
+	var count int
+	for scan.Next() {
+		if scan.Err() != nil {
+			err = scan.Err()
+			break
+		}
+		if err = dp(scan.Bytes()); err != nil {
```

Comment:
If this call to `dp` fails, is there any way to move the queue position backwards? Otherwise I'm thinking about a scenario where the writer is stuck retrying & never gets the write the data, and then the instance shuts down for some reason and when the queue restarts it is beyond that data, skipping the write.

Reply:
Calling `scan.Bytes()` won't advance the queue, it can be called as many times as needed until a call to `scan.Next()` is made. Its currently not set up to handle retries but once we hook in HTTP portions of this, it should be simple to retry here

Reply:
Resolved my concerns via a separate conversation - `scan.Next()` does _not_ advance the queue's on-disk position. That does not happen until `scan.Advance()` is called. So the scenario of a write waiting to retry and the instance restarting would result in, at works, a duplicated write attempt, which should not be a major concern.